flow/envs/multiagent/i210.py

```diff
@@ -156,53 +159,67 @@ def compute_reward(self, rl_actions, **kwargs):
-
-                    # weights for cost1, cost2, and cost3, respectively
-                    eta1, eta2 = 1.00, 0.10
-
-                    reward = max(eta1 * cost1 + eta2 * cost2, 0)
-
-                rewards[rl_id] = reward
+            speeds = self.k.vehicle.get_speed(self.k.vehicle.get_ids())
+            des_speed = self.env_params.additional_params["target_velocity"]
+            # rescale so the critic can estimate it quickly
+            reward = np.nan_to_num(np.mean([(des_speed - np.abs(speed - des_speed))**2
+                                              for speed in speeds]) / (des_speed**2))
+            rewards = {rl_id: reward for rl_id in self.k.vehicle.get_rl_ids()}
         return rewards
 
     def additional_command(self):
         """See parent class.
 
-        Define which vehicles are observed for visualization purposes.
+        Define which vehicles are observed for visualization purposes. Additionally, optionally reroute vehicles
+        back once they have exited.
         """
+        super().additional_command()
         # specify observed vehicles
         for rl_id in self.k.vehicle.get_rl_ids():
             # leader
             lead_id = self.k.vehicle.get_leader(rl_id)
             if lead_id:
                 self.k.vehicle.set_observed(lead_id)
 
+        if self.reroute_on_exit and self.time_counter >= self.env_params.sims_per_step * self.env_params.warmup_steps \
+                and not self.env_params.evaluate:
+            veh_ids = self.k.vehicle.get_ids()
+            edges = self.k.vehicle.get_edge(veh_ids)
+            for veh_id, edge in zip(veh_ids, edges):
+                if edge == "":
+                    continue
+                if edge[0] == ":":  # center edge
+                    continue
+                # on the exit edge, near the end, and is the vehicle furthest along
+                if edge == self.exit_edge and \
+                        (self.k.vehicle.get_position(veh_id) > self.k.network.edge_length(self.exit_edge) - 100)\
+                        and self.k.vehicle.get_leader(veh_id) is None:
+                    type_id = self.k.vehicle.get_type(veh_id)
+                    # remove the vehicle
+                    self.k.vehicle.remove(veh_id)
+                    lane = np.random.randint(low=0, high=self.num_enter_lanes)
+                    # reintroduce it at the start of the network
+                    # TODO(@evinitsky) select the lane and speed a bit more cleanly
+                    # Note, the position is 10 so you are not overlapping with the inflow car that is being removed.
+                    # this allows the vehicle to be immediately inserted.
+                    self.k.vehicle.add(
+                        veh_id=veh_id,
+                        edge=self.entrance_edge,
+                        type_id=str(type_id),
+                        lane=str(lane),
+                        pos="10.0",
+                        speed="23.0")
```

Comment:
Should this be desired_velocity or something?

Reply:
This is problematic but it just needs to match the inflow conditions.